openmc/material.py

```diff
@@ -641,37 +641,45 @@ def get_nuclide_atom_densities(self):
         nucs = []
         nuc_densities = []
         nuc_density_types = []
+
         for nuclide in nuclides.items():
             nuc, nuc_density, nuc_density_type = nuclide[1]
             nucs.append(nuc)
             nuc_densities.append(nuc_density)
             nuc_density_types.append(nuc_density_type)
 
+        nucs = np.array(nucs)
+        nuc_densities = np.array(nuc_densities)
+        nuc_density_types = np.array(nuc_density_types)
+
         if sum_density:
             density = np.sum(nuc_densities)
+
         percent_in_atom = np.all(nuc_density_types == 'ao')
         density_in_atom = density > 0.
         sum_percent = 0.
 
         awrs = []
-        for n, nuclide in enumerate(nuclides.items()):
+        for nuclide in nuclides.items():
             awr = openmc.data.atomic_mass(nuclide[0])
             if awr is not None:
-                awrs.append(awr / openmc.data.NEUTRON_MASS)
+                awrs.append(awr)
             else:
                 raise ValueError(nuclide[0] + " is invalid")
 
+        # Convert the weight amounts to atomic amounts
+        if not percent_in_atom:
+            for n, nuc in enumerate(nucs):
+                nuc_densities[n] *= self.average_molar_mass / awrs[n]
+
         # Now that we have the awr, lets finish calculating densities
         sum_percent = np.sum(nuc_densities)
         nuc_densities = nuc_densities / sum_percent
+
         if not density_in_atom:
-            sum_percent = 0.
-            for n, nuc in enumerate(nucs):
-                x = nuc_densities[n]
-                sum_percent += x * awrs[n]
-            sum_percent = 1. / sum_percent
-            density = -density * sum_percent * \
-                openmc.data.AVOGADRO / openmc.data.NEUTRON_MASS * 1.E-24
+            density = -density / self.average_molar_mass * 1.E-24 \
+                      * openmc.data.AVOGADRO
```

Comment:
This is converting the mass density (g/cm3) to atom density (atom/b-cm). First, the mass density (in g/cm3) is divided by the material average molar mass to get the molar density (moles/cm3). Then multiply by avogadro's number to get atom density (atoms/cm3). Finally multiply by 1e-24 to get density in units of atoms/barn-cm. I added a comment saying that we are converting mass density to atom density.